### ISSUE TITLE
[BUG] Migration Lock Row Not Removed After Migration

### ISSUE THREAD

**Issue by tomdeering-wf** [jobrunr/jobrunr#1139]
### JobRunr Version

7.3.1

### JDK Version

Corretto 21

### Your SQL / NoSQL database

Postgres 14

### What happened?

Dependabot recently upgraded our version of `org.jobrunr:jobrunr-pro-spring-boot-3-starter` from 7.2.3 to 7.3.0, and subsequently our service failed to deploy to several environments. Investigating, a prior migration run in July by version v7.2.2 ran a migration but failed to clean up its migration lock row (record with `id = 00000000-0000-0000-0000-000000000000` in `jobrunr_migrations`). Version 7.3.0 was also attempting to run a migration and failed because it could not acquire the migration lock.

![image](https://github.com/user-attachments/assets/78436616-8304-473f-8103-c393e1f930e1)

In order to unblock our service deploying with the new version of jobrunr we had to manually remove the lock row. It seems like there is an opportunity for jobrunr to be a more resilient so this does not occur again. This happened in several of our environments, so it seems likely to be a not-super-rare failure mode.

### How to reproduce?

Deploy service with jobrunr pro 7.2.2 using Postgres 14, observe that the table locker row is not cleaned up. If necessary, interrupt the service while jobrunr is migrating... I'm not sure whether this occurred in our environments.

### Relevant log output

```shell
exception: { [-]
     message: Database migrations have timed out.
     stacktrace: java.lang.IllegalStateException: Database migrations have timed out.
 at org.jobrunr.storage.sql.common.DatabaseCreator$MigrationsTableLocker.isMigrationsTableLocked(DatabaseCreator.java:422)
 at org.jobrunr.storage.sql.common.DatabaseCreator$MigrationsTableLocker.waitUntilMigrationsAreDone(DatabaseCreator.java:400)
 at org.jobrunr.storage.sql.common.DatabaseCreator$MigrationsTableLocker.access$300(DatabaseCreator.java:344)
 at org.jobrunr.storage.sql.common.DatabaseCreator.runMigrations(DatabaseCreator.java:164)
 at org.jobrunr.storage.sql.common.DatabaseCreator.runMigrations(DatabaseCreator.java:120)
 at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.setUpStorageProvider(DefaultSqlStorageProvider.java:89)
 at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.<init>(DefaultSqlStorageProvider.java:74)
 at org.jobrunr.storage.sql.common.DefaultSqlStorageProvider.<init>(DefaultSqlStorageProvider.java:61)
 at org.jobrunr.storage.sql.postgres.PostgresStorageProvider.<init>(PostgresStorageProvider.java:29)
 at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
 at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
 at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
 at org.jobrunr.storage.sql.common.SqlStorageProviderFactory.getStorageProvider(SqlStorageProviderFactory.java:77)
 at org.jobrunr.storage.sql.common.SqlStorageProviderFactory.getStorageProviderByJdbcUrl(SqlStorageProviderFactory.java:50)
 at org.jobrunr.storage.sql.common.SqlStorageProviderFactory.getStorageProviderUsingDataSource(SqlStorageProviderFactory.java:42)
 at org.jobrunr.storage.sql.common.SqlStorageProviderFactory.using(SqlStorageProviderFactory.java:36)
 at org.jobrunr.spring.autoconfigure.storage.JobRunrSqlStorageAutoConfiguration.sqlStorageProvider(JobRunrSqlStorageAutoConfiguration.java:39)
 at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
 at java.base/java.lang.reflect.Method.invoke(Method.java:580)
 at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:146)
 at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:644)
 at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:636)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1355)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1185)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:562)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
 at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
 at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
 at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
 at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
 at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
 at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1443)
 at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1353)
 at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:904)
 at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:782)
 at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:542)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1355)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1185)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:562)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
 at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
 at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
 at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
 at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
 at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
 at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1443)
 at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1353)
 at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:904)
 at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:782)
 at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:237)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1375)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1212)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:562)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:522)
 at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:337)
 at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
 at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:335)
 at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
 at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254)
 at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1443)
 at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1353)
 at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:904)
 at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:782)
 at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:542)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1355)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1185)
 at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCa
     type: java.lang.IllegalStateException
   }
   handler: IMgHAtp5
   host: { [+]
   }
   lang: 3
   level: error
   message: Error waiting for database migrations to finish. Manually review your database migrations in the jobrunr_migrations table and then delete the migration lock entry with id '00000000-0000-0000-0000-000000000000' before trying again.
   metadata: { [-]
     lineNumber: org.jobrunr.storage.sql.common.DatabaseCreator$MigrationsTableLocker.waitUntilMigrationsAreDone(DatabaseCreator.java:409)
     threadName: main
   }
```

**Comment by rdehuyss** [comment]
Hi @tomdeering-wf - the thing I wonder is why the lock was still there in the first place? You mention `interrupt the service while JobRunr is migrating`. In that case you're running in a partially migrated instance which is definitely not intended.

The lock is removed in a finally clause so only stays in the DB if for some reason your service is interrupted before JobRunr can finish the migrations. If that's the case, we exactly want you to review (hence the exception `Manually review your database migrations in the jobrunr_migrations table`) what has happened and we don't want to just continue tbh.

So it may be that the previous migrations did not complete successfully (e.g. missing indices) and this is something you should look into.

To us, this is behaving as expected and perhaps we even should not allow JobRunr to start if this row is there. 

Can you create a reproducer of this issue (without interrupting JobRunr during a migration)?

**Comment by tomdeering-wf** [comment]
@rdehuyss Thanks for the quick response! Unfortunately I don't have great repro steps for exactly what happened. Taking a look at one particular environment, it appears that the TABLE LOCKER record was inserted on 7/17/2024, but there wasn't actually a migration installed at that time (table contents attached as a CSV).

![image](https://github.com/user-attachments/assets/04a18bf0-d8be-4261-9d97-250673d0f5ba)

Your explanation makes sense that if jobrunr were truly interrupted in the middle of running migrations, a human really should review the database at that point. But upon review it doesn't _seem_ like that's what happened here.

[jobrunr_migrations.csv](https://github.com/user-attachments/files/17402330/jobrunr_migrations.csv)

**Comment by rdehuyss** [comment]
No worries, we're here to help... The reason we added the `MigrationsTableLocker` is that in deployments with multiple JobRunr servers, there may be cases where 2 servers try to run the migrations and the same time on startup. 

The `MigrationsTableLocker` prevents that by doing optimistic locking - only the server which can enter the TABLE LOCKER uuid (000...), can perform the actual migrations and other servers wait for them to be finished.

The actual locking and releasing is done in the [DatabaseCreator#runMigrations(List<SqlMigration> migrationsToRun)](https://github.com/jobrunr/jobrunr/blob/master/core/src/main/java/org/jobrunr/storage/sql/common/DatabaseCreator.java#L142-L158) and in a try/finally statement so I don't really see what could be happening there (unless the service is interrupted while doing the migration).

Feel free to have a look - a third / fourth pair of eyes would definitely be helpful!

**Comment by rdehuyss** [comment]
FYI - I just did the following migrations on Postgres 16.3 without problems:

- JobRunr Pro v7.2.3 (does all migrations up to v034)
- JobRunr Pro v7.3.0 (does migrations v035 and v036 like you mention but also removes the LOCK)
- JobRunr Pro v7.3.1 (no migrations)

**Comment by tomdeering-wf** [comment]
Ok, if I'm reading the flow of that code right (and keeping in mind that multiple servers will be executing this logic concurrently):

1. Jobrunr determines the appropriate migration scripts for the database in question and sorts them by name.
2. Without inserting the lock row, it separately queries to see whether each has been applied already, filtering that list down to only migrations that haven't yet been applied. If there are any:
3. Insert the lock row, which _should_ succeed for only one of the servers due to the PK constraint.
  a. For the server that got the lock
     i. Start a heartbeat process that will continue to update the timestamp of the lock row every 5 seconds.
     ii. Run each migration script in order.
     iii. Remove the migration lock row.
  b. For other servers that didn't get the lock
    i. Keep polling the migration lock row to see whether it's still there.
    ii. So long as the row is there and its timestamp isn't too old, continue waiting.
    iii. If the row's timestamp is too old, the server that had gotten the lock has been interrupted. Bail out. (this is the exception we were seeing)

I guess the mystery to solve would be, how was there a migration lock row that was inserted/last updated in July? That would require a server to think there was a new migration to run (there wasn't one at the time), insert that lock row, and then fail to clean it up. :thinking:

**Comment by tomdeering-wf** [comment]
Here's a thought. What if the SQLException [here](https://github.com/jobrunr/jobrunr/blob/v7.2.2/core/src/main/java/org/jobrunr/storage/sql/common/DatabaseCreator.java#L231), which is currently assumed to be because the table doesn't exist, was instead some other error (for example, soemthing transient due to networking)? Jobrunr would currently incorrectly conclude that the migration had not been applied yet.

**Comment by rdehuyss** [comment]
We just followed the code with your thought - even in that case, the lock should be removed.

Do you still have any logs from 7/17/2024? That may perhaps help us?

**Comment by tomdeering-wf** [comment]
Other than some telemetry showing that the service deployed at the time (makes sense), unfortunately there is nothing interesting in the logs. I do see that the infrastructure brought up a couple instances of the service for a very brief period of time, after which it killedl them off. That sometimes happens in Kubernetes environments as the infrastructure decides how best to schedule pods on the physical nodes that run them.

![image](https://github.com/user-attachments/assets/5b3a067d-12f7-43b0-9877-633cc41c5211)

Here's the only sequence of events that makes sense to me:

* Jobrunr misinterprets a SQLException, concluding that one or more migrations need to be applied.
* Jobrunr inserts the locker row
* The infrastructure decides to kill the instance of the service before it has a chance to execute its `finally` block that would remove the locker row.
* The fact that we have a locker row inserted remains a hidden issue that we don't notice. That only becomes apparent when a new version of jobrunr concludes in the future that there are new migrations to run, which didn't happen for months.

I can't prove this is what happened, but it's the only explanation that seems to fit the data. If that is what happened, there's not much jobrunr can do about the server being suddenly killed before that `finally` block executes. Other than better introspecting on that `SQLException`, I think the way jobrunr is applying its migrations makes sense!

**Comment by tomdeering-wf** [comment]
Maybe it would be possible to somehow hash/record the migrations jobrunr thinks it needs to run on that locker row? If that row encodes nothing, or only migrations that were already applied, and the timestamp is old, a new instance of the server could gracefully remove and ignore the row in this situation? In that case the presence of the locker row would not indicate partial application of migrations, right?

**Comment by rdehuyss** [comment]
Hi @tomdeering-wf, I like the idea you suggest but I prefer to find the root cause and I think there are too many assumptions for the moment.

In the current scenario's, we have 2 big ifs: 
- the `SQLException` (which is already done like that since the beginning of JobRunr) which means there would either be a network issue exception or a database problem? Although this can be the case, I've not seen that a lot (unless disks are full, ...).
- the fact that K8S decides to kill the POD so fast after starting
  - even in that case, there would be Threads interrupted that would also log something

Currently we log any exception when something goes wrong during the migrations... hence I was hoping to find something in your logs. I thought for a moment to register a shutdown hook but that would also not solve anything (as we don't know if the migrations were applied successfully in that case) so I'm out of idea's for the moment to be honest...

The only thing I can think of is not starting if the LOCK is there so this gets found earlier on?

**Comment by rdehuyss** [comment]
Hi @tomdeering-wf - we've decided that for now, we will not allow JobRunr to start if this row is still in the Migrations table.

One other idea: 
- could it be that at the time you started up (7/17/2024) there were troubles with the DB
- because of k8s healthcheck, the service did not came up within the given timeframe
- k8s decided to kill the pod keeping the row lock there?

**Comment by tomdeering-wf** [comment]
Hi @rdehuyss, I was out Thursday and Friday but am back today! 

> we've decided that for now, we will not allow JobRunr to start if this row is still in the Migrations table.

That makes sense to me!

> the SQLException (which is already done like that since the beginning of JobRunr) which means there would either be a network issue exception or a database problem? Although this can be the case, I've not seen that a lot (unless disks are full, ...).

There is a whole taxonomy of different kinds of [SQLException](https://docs.oracle.com/en/java/javase/21/docs/api/java.sql/java/sql/SQLException.html). As someone who has maintained a number of service backends, I can say I've seen many different kinds of SQLException for many different reasons. I would _strongly recommend_ not assuming the underlying reason that a `SQLException` encountered during migrations was thrown without further inspecting the specific exception. In your case I _think_ the kind of exception might be a [SQLSyntaxErrorException](https://docs.oracle.com/en/java/javase/21/docs/api/java.sql/java/sql/SQLSyntaxErrorException.html) for a query attempting to access a table that does not exist. 

> could it be that at the time you started up (7/17/2024) there were troubles with the DB
because of k8s healthcheck, the service did not came up within the given timeframe
k8s decided to kill the pod keeping the row lock there?

Yes, this is definitely possible. In general it is always possible for infrastructure to kill a server without giving it a chance to exit gracefully. Graceful shutdown is best-effort and isn't guaranteed. I can't say for sure why those pods were killed at the time; there are many possible reasons, including what you suggest.

**Comment by tomdeering-wf** [comment]
> That makes sense to me! This at least would have allowed us to discover the issue at the time it happened rather than months later.

I'm thinking through this some more. Refusing to start Jobrunr if the migration lock row is still present, regardless of whether Jobrunr thinks it has any migrations to run, could make sense. It would have immediately revealed the problem to us, but it also would have put us in a service outage situation where reverting to a previous release or older version of jobrunr would not help. In other words, it would have increased the severity of this incident for us. 

Have you thought through that angle @rdehuyss?

**Comment by rdehuyss** [comment]
Hi @tomdeering-wf - I don't understand how downtime after the previous deploy would be any different after the downtime you now had due to the LOCK row still being there?

I think it should still be visible as soon as possible and not months later so we're adding this guardrail.

**Comment by tomdeering-wf** [comment]
Hopefully this helps clarify what I mean.

**Previous:**
* Server V1 inserted the migration locker row (we're still unsure why, although misinterpreting a SQLException seems a likely explanation) then failed to clean it up. This was "ok" because there weren't any migrations that failed to apply and the server could still do its job.
* Server V2 failed on deploy because it had migrations to run and couldn't. Server V1 remained deployed, no application downtime. 
* Deploying Server V2+ was blocked until we manually removed the lock row or reverted jobrunr to v7.2.2. Not great, but at no point was there application downtime.

**Proposed:**
* Server V1 inserts the migration locker row and fails to clean it up. Server V1 refuses to start. Any time the infrastructure moves a Server V0 container to a new node, that too observes the lock row and fails to start. Application downtime begins.
* Trying to revert to Server V0 doesn't help the situation. Downtime continues.
* Downtime continues until the lock row is manually removed.

**Comment by rdehuyss** [comment]
Hi @tomdeering-wf - well, with the extra precondition that we now can't misinterpret the SQLException, I think this is ok.

And I think your 2 scenario's are not comparable or I don't understand your deployment scenario exactly (as I read you're doing rolling updates):

## How I see it (using rolling update):
- Server v1 is running with previous version
- Server v2 tries to deploy new version but fails and keeps lock in DB. It will not start until deploys are verified. Server v1 continues running as once it is running, it will not check the DB for the locking row.

<hr>

The [proposed solution](https://github.com/jobrunr/jobrunr/pull/1142) adds a lot of benefit:
- `SQLException`s will not be treated anymore as before - if there is one, it is an actual problem with the DB and the deploy stops. It solves one of our assumptions (that there was an `SQLException` unrelated to the missing table).
- We will see any issues related to the table locking a lot earlier.

Last but not least I hope you understand that the time we're spending on this corner-case issue is getting out of hand. We will merge the PR and if you really want to be sure, you can monitor the lock as part of your deployment.

**Comment by rdehuyss** [comment]
Solved by PR https://github.com/jobrunr/jobrunr/pull/1142.
